### PR TITLE
Features/bp 266 smart enums

### PR DESCRIPTION
**Pull request by zsrdjan**

### CODE REVIEW

File: Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/ApiSetupFixture.GivenValidUserId_GetById_ReturnsUserDetails.verified.txt [hunk 1]
```diff
@@ -4,5 +4,11 @@
   FullName: John Doe,
   RoleId: Guid_2,
   CreatedOn: DateTimeOffset_1,
-  UpdatedOn: DateTimeOffset_1
+  UpdatedOn: DateTimeOffset_1,
+  UserStatusId: {
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/ApiSetupFixture.GivenValidUserId_GetById_ReturnsUserDetails.verified.txt:8]: Smart enum

File: Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/ApiSetupFixture.cs [hunk 2]
```diff
@@ -25,7 +25,8 @@ public void SetUp()
         _user = new UserBuilder()
             .WithEmailAddress("john_doe@john.doe")
             .WithFullName("John Doe")
-            .WithRoleId(Role.SystemAdminRoleId);
+            .WithRoleId(Role.SystemAdminRoleId)
+            .WithStatusId(UserStatusId.Active);
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/ApiSetupFixture.cs:29]: Smart enum

File: Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/ApiSetupFixture.cs [hunk 3]
```diff
@@ -85,9 +100,23 @@ public async Task TestCreateReturnsValidationErrors(string name,
...
+            RoleId = Role.SystemAdminRoleId,
+            UserStatusId = UserStatusId.Inactive
+        };
+        var response = await Client.PostAsJsonAsync("users", command, HttpSerializationOptions.Options);
 
         response.Should().BeBadRequest().And.ContainValidationError(validationField, validationErrorMessage);
     }
+
+    [Test]
+    public void TestAutoMapperMappings()
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/ApiSetupFixture.cs:117]: I have noticed that there was a problem with Automapper mappings (ProductController). I have added test that checks mappings.

File: Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/AuthorizationFixture.cs [hunk 4]
```diff
@@ -60,10 +59,8 @@ public async Task EndpointWithoutAuthorizeAttributeIsAllowed()
     public async Task TearDown()
     {
         var currentUser = QueryCurrentUser();
-        currentUser.Update(new CreateOrUpdateUser.Command
-        {
-            RoleId = Role.SystemAdminRoleId
-        });
+        var role = QueryDb<Role>().First(r => r.Id == Role.SystemAdminRoleId);
+        currentUser.UpdateRole(role);
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/CoreFeatures/AuthorizationFixture.cs:63]: Simplify update role. Command was not properly created.

File: Enigmatry.Entry.Blueprint.Api.Tests/Features/UsersControllerFixture.cs [hunk 5]
```diff
@@ -49,10 +51,33 @@ public async Task TestGetById()
         await Verify(user);
     }
 
+    [Test]
+    public async Task TestGetRoles()
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/Features/UsersControllerFixture.cs:55]: Test was missing

File: Enigmatry.Entry.Blueprint.Api.Tests/Features/UsersControllerFixture.cs [hunk 6]
```diff
@@ -49,10 +51,33 @@ public async Task TestGetById()
...
+    {
+        var user = await Client.GetAsync<IEnumerable<LookupResponse<Guid>>>($"users/roles");
+
+        await Verify(user);
+    }
+
+    [Test]
+    public async Task TestGetStatuses()
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/Features/UsersControllerFixture.cs:63]: New functionality introduced

File: Enigmatry.Entry.Blueprint.Api.Tests/Features/UsersControllerFixture.TestUpdate.verified.txt [hunk 7]
```diff
@@ -4,5 +4,8 @@
   FullName: some user,
   RoleId: Guid_2,
   CreatedOn: DateTimeOffset_1,
-  UpdatedOn: DateTimeOffset_1
+  UpdatedOn: DateTimeOffset_1,
+  UserStatusId: Inactive,
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/Features/UsersControllerFixture.TestUpdate.verified.txt:8]: UserStatusId: Inactive was achieved with Enigmatry.Entry.SmartEnums.VerifyTestsJson.ConvertersExtensions.EntryAddSmartEnumJsonConverters.

If this was not registered the Verify would contain both the Id and the name in the UserStatusId which makes reading Verify files a bit more dificult and confusing.

File: Enigmatry.Entry.Blueprint.Api.Tests/Infrastructure/ApiWebApplicationFactory.cs [hunk 8]
```diff
@@ -59,8 +58,9 @@ protected override IHost CreateHost(IHostBuilder builder)
         return base.CreateHost(builder);
     }
 
-    private static void ConfigureContainer(ContainerBuilder builder) =>
-
-        // in the api tests we need to replace current user with TestUser
-        builder.RegisterModule(new TestModule(true)); // this allows certain components to be overriden
+    private static void ConfigureContainer(ContainerBuilder builder)
+    {
+        builder.RegisterModule<IdentityModule<TestUserProvider>>();
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api.Tests/Infrastructure/ApiWebApplicationFactory.cs:63]: Fixes another problem in the blueprint (i.e. resolving CurrenUser in the tests) which was blocking me from fixing the tests.

File: Enigmatry.Entry.Blueprint.Api/Features/Users/GetUserDetails.cs [hunk 9]
```diff
@@ -26,35 +28,32 @@ public class Response
         public Guid RoleId { get; set; }
         public DateTimeOffset CreatedOn { get; set; }
         public DateTimeOffset UpdatedOn { get; set; }
+        public UserStatusId UserStatusId { get; set; } = UserStatusId.Active;
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Api/Features/Users/GetUserDetails.cs:31]: Smart enum.

File: Enigmatry.Entry.Blueprint.ApplicationServices/Auditing/AuditableDomainEventHandler.cs [hunk 10]
```diff
@@ -7,21 +7,15 @@
...
-    public AuditableDomainEventNotificationHandler(ILogger<AuditableDomainEventNotificationHandler<T>> log, ICurrentUserProvider currentUserProvider)
-    {
-        _log = log;
-        _currentUserProvider = currentUserProvider;
-    }
-
     public Task Handle(T notification, CancellationToken cancellationToken)
     {
         // here you can enter record in Audit table, 
-        _log.LogDebug("Event name: {EventName}, Payload: {@Payload}, initiated by: {EmailAddress}", notification.EventName, notification.AuditPayload, _currentUserProvider.User?.EmailAddress);
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.ApplicationServices/Auditing/AuditableDomainEventHandler.cs:24]: We should not log EmailAddress

File: Enigmatry.Entry.Blueprint.Common.Tests/CommonVerifierSettingsInitializer.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Runtime.CompilerServices;
+using Argon;
+using Enigmatry.Entry.Blueprint.Infrastructure.Api.Init;
+using Enigmatry.Entry.SmartEnums.VerifyTests;
+
+namespace Enigmatry.Entry.Blueprint.Common.Tests;
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Common.Tests/CommonVerifierSettingsInitializer.cs:6]: New project (Common.Tests). Idea is to move common test related things from Infrastructure here (e.g. Builders). We keep adding this project later.

File: Enigmatry.Entry.Blueprint.Data.Migrations/DbInitializer.cs [hunk 12]
```diff
@@ -5,7 +5,12 @@ namespace Enigmatry.Entry.Blueprint.Data.Migrations;
 
 public static class DbInitializer
 {
-    private static readonly List<ISeeding> Seeding = [new RolePermissionSeeding(), new UserSeeding()];
+    private static readonly List<ISeeding> Seeding =
+    [
+        new UserStatusSeeding(),
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Data.Migrations/DbInitializer.cs:10]: Seeding of SmartEnum

File: Enigmatry.Entry.Blueprint.Data.Migrations/Seeding/UserStatusSeeding.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿using Enigmatry.Entry.Blueprint.Domain.Users;
+
+namespace Enigmatry.Entry.Blueprint.Data.Migrations.Seeding;
+
+internal class UserStatusSeeding() : EntityWithEnumIdSeeding<UserStatus, UserStatusId>(statusId =>
+    new { Id = statusId, statusId.Name, Description = GetDescription(statusId) })
+{
+    private static string GetDescription(UserStatusId statusId) =>
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Data.Migrations/Seeding/UserStatusSeeding.cs:8]: Description field was added to showcase how to populate extra column for each given enum value.

File: Enigmatry.Entry.Blueprint.Domain.Tests/Enigmatry.Entry.Blueprint.Domain.Tests.csproj [hunk 14]
```diff
@@ -5,16 +5,15 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="FluentAssertions" />
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Domain.Tests/Enigmatry.Entry.Blueprint.Domain.Tests.csproj:8]: Moved to common

File: Enigmatry.Entry.Blueprint.Domain.Tests/Enigmatry.Entry.Blueprint.Domain.Tests.csproj [hunk 15]
```diff
@@ -5,16 +5,15 @@
...
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
     <PackageReference Include="NUnit3TestAdapter">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="Verify.NUnit" />
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Domain.Tests/Enigmatry.Entry.Blueprint.Domain.Tests.csproj:14]: Moved to common

File: Enigmatry.Entry.Blueprint.Domain/Identity/UserQueryableExtensions.cs [hunk 16]
```diff
@@ -1,9 +1,15 @@
 ﻿using Enigmatry.Entry.Blueprint.Domain.Users;
+using Microsoft.EntityFrameworkCore;
 
 namespace Enigmatry.Entry.Blueprint.Domain.Identity;
 
 public static class UserQueryableExtensions
 {
+    public static IQueryable<User> BuildAggregateInclude(this IQueryable<User> query) =>
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Domain/Identity/UserQueryableExtensions.cs:8]: On Windesheim we use this naming convention in cases where we need to load entity with its related children to make root aggregate complete (this is usually used for the updates, not queries)

File: Enigmatry.Entry.Blueprint.Domain/Users/UserStatus.cs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+using Enigmatry.Entry.Blueprint.Core.Entities;
+using JetBrains.Annotations;
+
+namespace Enigmatry.Entry.Blueprint.Domain.Users;
+
+[UsedImplicitly(ImplicitUseTargetFlags.WithMembers)]
+public class UserStatus : EntityWithEnumId<UserStatusId>
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Domain/Users/UserStatus.cs:7]: Smart enum related table

File: Enigmatry.Entry.Blueprint.Domain/Users/UserStatus.cs [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+{
+    public string Description { get; private set; } = String.Empty;
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Domain/Users/UserStatus.cs:9]: With an additional property that needs to be stored in the db.

File: Enigmatry.Entry.Blueprint.Infrastructure.Tests/Autofac/TestModule.cs [hunk 19]
```diff
@@ -1,35 +1,15 @@
...
-        }
-
+    protected override void Load(ContainerBuilder builder) =>
         builder.RegisterType<SettableTimeProvider>()
             .As<ITimeProvider>()
             .AsSelf()
             .SingleInstance();
-    }
-
-    private static void ReplaceCurrentUserWithTestUser(ContainerBuilder builder) =>
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Infrastructure.Tests/Autofac/TestModule.cs:33]: No longer needed. Fixed with TestCurrentUserProvider.

File: Enigmatry.Entry.Blueprint.Infrastructure/Data/PopulateCreatedUpdatedInterceptor.cs [hunk 20]
```diff
@@ -36,21 +36,15 @@ public override ValueTask<InterceptionResult<int>> SavingChangesAsync(DbContextE
 
     private void PopulateCreatedUpdated(DbContext dbContext)
     {
-        var currentUser = _currentUserProvider().User;
-        if (currentUser == null)
-        {
-            // TODO: Discuss throwing an exception here; Currently seeding data in the integration tests would fail because of it
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Infrastructure/Data/PopulateCreatedUpdatedInterceptor.cs:42]: This was a problematic solution and it was now fixed with SystemUserProvider and TestUserProvider.

File: Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+using Enigmatry.Entry.Blueprint.Domain.Users;
+using Enigmatry.Entry.Core.Data;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+
+namespace Enigmatry.Entry.Blueprint.Infrastructure.Identity;
+
+[UsedImplicitly]
+public class PrincipalProviderCurrentUserProvider(
```
Comment (zsrdjan) [inline Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: Renamed class CurrentUserProvider

File: nuget.config [hunk 22]
```diff
@@ -3,6 +3,7 @@
 <configuration>
   <packageSources>
     <clear/>
+    <add key="LocalPackages" value="c:\Projects\local_nugets\" />
```
Comment (zsrdjan) [inline nuget.config:6]: This is temp. Will be removed as soon as Nugets have been published.

File: Enigmatry.Entry.Blueprint.Domain/Identity/UserContext.cs [hunk 23]
```diff
@@ -0,0 +1,7 @@
+namespace Enigmatry.Entry.Blueprint.Domain.Identity;
+
+public class UserContext(Guid id, PermissionsContext permissions)
+{
+    public Guid Id => id;
```
Comment (edekoning) [inline Enigmatry.Entry.Blueprint.Domain/Identity/UserContext.cs:5]: Shouldn't it be UserId? Or is that too obvious?

File: Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs (hunk 21 above)
Comment (zsrdjan) [reply Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: This was done to make explicit that this implementation has a dependency in IPrincipal (i.e. it requires authentication and claims to be setup, which makes it more difficult to use in the scheduler or tests).  

This is the fix that removed the workaround in the PopulateCreatedUpdateInterceptor.cs. PopulateCreatedUpdateInterceptor used to check whether UserId was not null and exit early in that situation leaving CreatedById and UpdatedById to have Guid.Empty which caused problems in the tests seeding.
 
Originally, we have had three implementations of ICurrentUserProvider: 
- CurrentUserProvider (that has dependency on IClaimsProvider which has dependency on IPrincipal and ClaimsPrincipal (ClaimsPrincipal cannot be resolve without setting up the Authentication system)
- NullCurrentUserProvider - the class is no longer being used so it was removed
- SystemUserProvider - had a hard-code UserId
In our tests we are seeding test user (under which the tests will be executed in the api) but none of the implementations allow id of the user to be change to test user. 
The fix to remove the workaround was to introduce TestUserProvider (which inherits from SystemUserProvider but overrides the userId). During that fix, I removed NullCurrentUserProvider and renamed CurrentUserProvider to make its dependency more explicit. 

If you think that it is better to rename it back to CurrentUserProvider that is also fine. I don't have any strong preference here.

File: Enigmatry.Entry.Blueprint.Domain/Identity/UserContext.cs (hunk 23 above)
Comment (zsrdjan) [reply Enigmatry.Entry.Blueprint.Domain/Identity/UserContext.cs:5]: I will rename it to UserId, it makes sense.

File: Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs (hunk 21 above)
Comment (zsrdjan) [reply Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: This is what is seen in the tools:
image

Comment (zsrdjan) [reply Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: My idea was to make this more similar to the "System" and "Test"

Comment (edekoning) [reply Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: Yeah, I understand your reason. It's just that most devs are used to just 'CurrentUserProvder'. And from a naming perspective... it does provide the current user, while SystemUserProvider provides a system user.

Comment (edekoning) [reply Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: TestUserProvider could IMO also be TestCurrentUserProvider. Both make sense.

Comment (zsrdjan) [reply Enigmatry.Entry.Blueprint.Infrastructure/Identity/PrincipalProviderCurrentUserProvider.cs:11]: Ok, your point also makes sense, I will then consolidate into CurrentUserProvider, SystemUserProvider and TestUserProvider

**Review by edekoning — APPROVED** [review]

**Pull request merged**